### ISSUE TITLE
Multimap becomes inconsistent when remove before shutdown

### ISSUE THREAD

**Issue by purplefox** [hazelcast/hazelcast#5220]
Hello folks,

We're experiencing a rather serious issue in Vert.x which results in inconsistent/corrupt multimap entries in the hazelcast cluster.

I have managed to reproduce this using pure Hazelcast APIs (no Vert.x involved).

Reproducer is here https://github.com/purplefox/hz-bug

The README contains a full explanation and instructions for reproducing.

There is also a thread from the Vert.x google group discussing this here:

https://groups.google.com/forum/?fromgroups#!topic/vertx/Im4UumOG8Ts

This issue is currently affecting Vert.x users in production, so if you could take a look / provide a workaround, that would be most helpful.

**Comment by gurbuzali** [comment]
thank you for the detailed bug report, we'll take a look at it

**Comment by gurbuzali** [comment]
Are you sure that all `remove` calls return true ?
because in my tests I've checked and counted `false` returns and it was equal to the remaining entry count 
(node1: 21, node2: 31, node3: 0, node4: 0, remaining entries: 52)

**Comment by purplefox** [comment]
Yes, you're right - sometimes the remove fails.

However, why should the remove fail? It's just removing exactly the same key, value that it previously put in the map, and no other node has removed it.

**Comment by gurbuzali** [comment]
if you shutdown more than 1 node concurrently, operations (in this case remove operation) may get timeout due to migration. but in that case we should've seen some timeout exceptions I guess. 
That's another issue I will look in more detail, let's confirm that remaining entry count is equal to "not removed entry count". Do you see a case which these do not match ? if so we will have 2 different issues I presume.

**Comment by purplefox** [comment]
I think the issue here is that it's failing to remove entries which are clearly in the map.

IMHO it's a fundamental characteristic of any clustered/distributed map solution that the map should remain consistent when nodes are shutdown as long as there is at least one node in the cluster.

If a node shuts down and owns certain keys I'd expect that it must ensure those keys are migrated safely to another node before shutdown is complete. I'm guessing that isn't happening properly in this case.

I'm not seeing any timeouts, everything happens pretty quickly.

**Comment by gurbuzali** [comment]
I'm not saying anything against your argument, I'm just trying to pinpoint the issue here.
Do you confirm that the number of "false returns" is equal to remaining entries ?

**Comment by purplefox** [comment]
Ah ok :)

I thought that your statement "may get timeout due to migration" implied that this was to be expected, and I assume this would leave the cluster in an inconsistent state, which clearly would be a major issue - users can shutdown nodes at any time and that's not something we can control.

I don't know if the number of false returns is equal to the number of remaining entries, but that could well be true. If you run it a few times maybe you could get some degree of confidence that is the case?

**Comment by purplefox** [comment]
Btw.. thanks for looking at this so quickly, it's much appreciated :)

**Comment by bwzhang2011** [comment]
@purplefox, thanks for such issue. I wonder such was some actual bug need to be fixed or some that would exist under some circumstance that we need pay attention to be prevented from.

**Comment by AlexThurston** [comment]
This was an actual bug encountered by myself when trying to use Vertx which in turn uses Hazelcast that effectively rendered our clustered system to be non-functional.

Others have noted similar behaviour.

**Comment by AlexThurston** [comment]
Is there any movement/decision on this issue?  Just trying to plan and mitigate accordingly.

**Comment by purplefox** [comment]
+1 This is very important for our users. If there is any sensible workaround that can be implemented before a full solution is provided, that would be useful too :)

**Comment by ajermakovics** [comment]
One workaround is to use an IMap instead of a MultiMap. I ran a [modified test](https://github.com/ajermakovics/hz-bug/commit/8512237c314a9692ae13331dfbb8f3959c4321aa) and all the removes succeeded. Currently the IMap implementation is more mature than the MultiMap.

Of course IMap doesn't support multiple values per key but you could use multiple maps. If the names of these IMaps share a common prefix then they can also share configuration.

**Comment by purplefox** [comment]
Thanks @ajermakovics for that observation. 

Perhaps it will provide a clue to the Hazelcast team about where the bug is...

**Comment by jensenmo** [comment]
+1 from us; we just had the issue in production again yesterday - only known workaround is to bounce the whole vert.x/hazelcast cluster, which causes an outage

**Comment by purplefox** [comment]
Hi Hazelcast folks

just wondering if there has been any progress on this or if you know if any workaround?

We have a code freeze for Vert.x 3.0 in the not too distant future so would need some kind of solution before then.

I know these things are hard to tell but if you could give some kind of ETA for this it would be much appreciated, as we will need to take mitigating actions if it looks like its not going to be fixed ready for Vert.x 3.0.

Thanks :)

**Comment by lzeligowski** [comment]
+1 from my team. We've got the same problems as other (vert.x cluster). From our perspective - this is critical error.

**Comment by gurbuzali** [comment]
Hi
I've sent a PR #5284 , once it's merged can you please try and confirm if it fixes the issue?

**Comment by purplefox** [comment]
Thanks Ali,

I have built hz from your fix branch and I can confirm that my reproducer seems to work as expected now.

I wonder if you could confirm one thing for me? Looking at your fix.. when the multimap values are merged you are setting the new id sequence to be the max of the merged ids plus an offset (=1000000). I am guessing this offset exists to account for any puts that might occur during the merge process (?). Is there a possibility if many puts (more than 1000000) occurred during the merge then we would still have problems?

**Comment by purplefox** [comment]
@AlexThurston Could you also try with your reproducer and Hz built from Ali's fix branch here: https://github.com/gurbuzali/hazelcast/tree/multimap-migration-fix ?

**Comment by purplefox** [comment]
@gurbuzali Could you also confirm when the next Hazelcast release is due on the 3.4 branch? (I.e. the one containing this fix).

**Comment by gurbuzali** [comment]
Hi @purplefox while merging all mutating operations retried so there may not be a put while merging. Problem here is with the transactions. Transactions gets nextRecordId but may not be committed at the time of migration. So if you have more that 1000000 transactions alive on multimap while migrating then you may have problems. We believe this is a very unlikely situation.
We use the same approach in IQueue too

PS: By transaction I mean using our `TransactionalMultiMap` interface

**Comment by gurbuzali** [comment]
For the next maintenance release (3.4.3) I don't have an exact date, I will ask our release coordinator and come back to you.

**Comment by AlexThurston** [comment]
@purplefox I have good news and bad news.  The good news is that the behaviour has changed and for the better.  I am testing through Vertx which was the original impetus of this investigation.

I'll speak in mostly Vertx terms, and perhaps @purplefox can investigate if this is still related to only Hazelcast.

If I have 3 vertx process which are running verticles that register 100 handlers each, the map's size is correctly 300 (3 * 100) when everything is running.

I then (in parallel) kill 2 of the three vertx process.  The resulting map correctly only contains entires for the first vertx process.  **However**, there are handlers missing.  I should be left with a map of size 100, but have a map of size 73 instead.  Basically, I've lost handlers 27 handlers by shutting down Vertx processes.

Starting the two Vertx process that were previously killed results in a map that does not correctly reflect what is actually running in the cluster.

This is still a pretty serious and breaking issue (although slightly different), but I can't say if it's Vertx of Hazelcast that's the cause.

**Comment by purplefox** [comment]
@AlexThurston can you clarify the exact reproducer you are using - I assume its this one here? https://github.com/purplefox/vertx_bug/tree/tim_reproducer

**Comment by AlexThurston** [comment]
@purplefox Yup.  I just used the code there to demonstrate the problem as well.  Had to change things around slightly to make sure I was using the newly built Hazelcast jars, but it demonstrates the problem.

At the end of the script, once things have been killed, the check shows a map that only contains servers running on port 6666 as it should, but the map is only 39 elements long when it should be 50.

**Comment by purplefox** [comment]
@AlexThurston OK, let's continue on the google group - can you provide your output?

**Comment by gurbuzali** [comment]
so is it related with hazelcast or not? :)

**Comment by purplefox** [comment]
@gurbuzali I'm not sure if it's the same bug or a different one, but I have created a new reproducer here:

https://github.com/purplefox/hz-bug/tree/second_bug

Which shows an issue when using 3.4.3-SNAPSHOT version that includes your latest fix.

This reproducer is a bit simpler to run than the last one.

Just run the Runner class in your IDE. You may have to run it several times in order to see the issue.
When the issue occurs you should see

Map Size: 35

displayed in the console.

There's some more info in the README on what the reproducer does.

Thanks.

**Comment by purplefox** [comment]
Hi @gurbuzali , did you see the new reproducer I added above?

Do you think it is the same issue or a different one? I am happy to open a new bug report for that if you prefer.

**Comment by gurbuzali** [comment]
Hi @purplefox 
I've reproduced the issue yesterday but couldn't investigate it more deeply.
My guess is that this is a different issue which may be related to internals not multimap.
I'll investigate more

**Comment by AlexThurston** [comment]
@gurbuzali This is great news.  Looking forward to hearing what you've found (and even more to the fix) :smile:

**Comment by AlexThurston** [comment]
@gurbuzali Any progress nailing this down?

Sorry for bugging ya, I'm keen :smile:

**Comment by purplefox** [comment]
@gurbuzali +1 Any updates much appreciated :)

**Comment by bwzhang2011** [comment]
@gurbuzali, any update with such issue.

**Comment by purplefox** [comment]
@gurbuzali Could you give us an update on the progress? Even if it's to say "I am still working on it and hopefully will have a solution in the next couple of days", is better than not saying anything at all.......

**Comment by gurbuzali** [comment]
@purplefox sorry for not updating the issue, was a bit busy with some other stuff.
For the issue, I don't have anything new to say. I will give it more time this weekend .

**Comment by gurbuzali** [comment]
Hi @purplefox short update, looks like the issue is not related with multimap but the handling of concurrent shutdown. We don't have a verified fix yet, still investigating for one.

**Comment by bwzhang2011** [comment]
@gurbuzali, maybe @pveentjer  or  @noctarius  could leave us some advice.  I wonder whether hz3.5 has the same issue.

**Comment by bwzhang2011** [comment]
@purplefox, did you test it with the latest hazelcast (hz3.5 snapshot)

**Comment by jerrinot** [comment]
Hi,
we fully understand the root cause now. There is a race in a concurrent shutdown process. It's safe to shutdown at most X members where X is no. of sync backups. 

I saw your workaround https://github.com/eclipse/vert.x/commit/def859b5ec98c388aafc937035d8087fbd0cd99b I quite like it; it's a good idea!

Improvements in the graceful shutdown procedure are on our radar for a future Hazelcast version.

**Comment by purplefox** [comment]
Hi @jerrinot could you elaborate a bit more on "It's safe to shutdown at most X members where X is no. of sync backups." ?

Do you mean that you cannot guarantee the consistency of the data if more than X members are *concurrently* shutdown where X is number of sync backups?

**Comment by jerrinot** [comment]
@purplefox: that's correct. there is a tiny chance of data-loss in the case.

**Comment by purplefox** [comment]
Can you quantify "tiny chance"?

Some of our users can easily reproduce this in a real set-up by shutting down servers concurrently resulting in an inconsistent state. I.e. it doesn't appear to be a tiny chance, there appears to be a very significant probability of it happening. This can be seen in the reproducer I provided where the cluster gets into an inconsistent state in the majority of runs, so I'd say around a 75% chance of it happening in this setup.

**Comment by jerrinot** [comment]
Here is what's going on:

Let's assume we have a cluster of 3 members: Member1, Member2, Member3 and we have just 3 partitions in a system.

```
Partition Table:

           / Member1 | Member2 | Member 3
Primary of |   0     |    1    |    2
Backup  of |   2     |    0    |    1
```

During a shutdown `Member1` checks if `Member2` has an update-to-date replica of a `partition0` and it will exit only if the outcome is positive. Otherwise the `Member1` waits until the backup is up-to-date. 

This is the usual flow:
```
- Member1 -> Member2: Is your replica of `partition0` at version X
- Member2 -> Member1: Yes
- Member1 exits
- Member2 is promoted to be a new owner of `partition0`
- Member3 becomes a backup of `partition0`
```

However when `Member2` calls `shutdown` before it's promoted to be a new owner, but after `Member1` is killed then you will lose `partition0`

I was considering a quick workaround: If a member is in a process of shutting down then it will always reply `NO` to the "`IsYourReplicateUpToDate`" question, but:
1. it could cause dead-locks. 
2. it doesn't solve all cases anyway

The shutdown process has to be re-designed. It's very likely to be a part of the 3.6 release.

**Comment by purplefox** [comment]
OK, thanks for the update.

**Comment by jerrinot** [comment]
No problem at all. I understand the pain caused by the current situation. I'll do my best to have a proper fix in 3.6. 

Feel free to contact me if you have any question / workaround to discuss, etc.

**Comment by bwzhang2011** [comment]
@jerrinot, there's a doubt that shutdown logic is hidden behind as hz3.5 final has not been released and why it will be postponed to hz3.6. could it not be backport into 3.4.4 or hz3.5.1 ?

**Comment by jerrinot** [comment]
@bwzhang2011: It's hard to answer it right now as I don't know how the final fix will look like at this moment. We tend to backports things only when we are very pretty sure they don't have any side-effects.

We'll update this ticket once we start designing the new shutdown process.

**Comment by jaysonminard** [comment]
Any updated @jerrinot

**Comment by jaysonminard** [comment]
or @purplefox , this looks deadly to vert.x 3 clustering

**Comment by jerrinot** [comment]
hi @jaysonminard not really. It's something we'll likely address in Hazelcast 3.6. I believe I saw a workaround somewhere in the vert.x codebase.

**Comment by purplefox** [comment]
Yes it's all a bit of a mess. Luckily we have a workaround for this one which seems reasonable but we also need the following https://github.com/hazelcast/hazelcast/pull/5284#issuecomment-106797517 fixed in a release in the next few days.

**Comment by purplefox** [comment]
Yes we have a workaround for this but we need 5285 fixed too (above) and we don't have a workaround for that.

**Comment by purplefox** [comment]
If the worst comes to the worst we will have to fork and make our own hazelcast release that provides the fixes which isn't ideal, but not the end of the world.

Clearly it would be preferable if Hazelcast implemented the fixes themselves!

**Comment by jerrinot** [comment]
yeah, I understand the pain. 
I'm not sure what the status of 3.4.3 is - most of our devs including myself are busy with last changes for 3.5 these days. 

I'll trying to get a timeline for 3.4.3 and I'll update this ticket once I'll get some info.

**Comment by jerrinot** [comment]
Hi, I talked to my colleagues and we are aiming to release 3.4.3 by the end of this week. This assumes all our internal tests will pass.

**Comment by purplefox** [comment]
Ok, thanks @jerrinot please keep us posted.

**Comment by jaysonminard** [comment]
yay @jerrinot

**Comment by jerrinot** [comment]
Hi guys,
I apologize for not being very responsive. Our internals tests for 3.4.3 just passed. You can expect the release in no more than a few days.

**Comment by jerrinot** [comment]
Hazelcast 3.4.3 was just released.

**Comment by jaysonminard** [comment]
yay!

**Comment by jerrinot** [comment]
yay indeed!

**Comment by jaysonminard** [comment]
@purplefox can you please bump version of hazelcast for vert.x-3 and do a milestone7?

**Comment by jensenmo** [comment]
@purplefox I would second that - but for 2.1.x? Thanks!

**Comment by jaysonminard** [comment]
(its a good sign that people fight over wanting vert-x releases, because vert-x is so loved jaja)

**Comment by purplefox** [comment]
@jerrinot @gurbuzali Unfortunately the 3.4.3 release is broken and not usable by Vert.x:

https://github.com/hazelcast/hazelcast/issues/5538

This is becoming extremely critical for us. We have a major release of Vert.x due next Monday, and right now we don't have a stable version of Hazelcast that we can use for it.

I hope the hazelcast team will look at this very urgently.
